### ISSUE TITLE
xtigervncviewer returns 0 even though there was a connection error

### ISSUE THREAD

**Issue by finefoot** [TigerVNC/tigervnc#1008]
If there is no VNC server running on localhost and I still try to connect, I get the following output:

```none
$ xtigervncviewer -AlertOnFatalError=0 localhost

TigerVNC Viewer 64-bit v1.9.0
Built on: 2020-01-23 18:03
Copyright (C) 1999-2018 TigerVNC Team and many others (see README.rst)
See http://www.tigervnc.org for information on TigerVNC.

Fri May  1 15:55:44 2020
 DecodeManager: Detected 4 CPU core(s)
 DecodeManager: Creating 4 decoder thread(s)
 CConn:       unable to create socket: Address family not supported by protocol
              (97)
$ echo $?
0
```

I would have expected a non-zero exit code.

**Comment by CendioOssman** [comment]
We don't really use exit codes right now, but I suppose we could do so in the future.

**Comment by finefoot** [comment]
> We don't really use exit codes right now

For example, if I run `xtigervncviewer` and click on <kbd>Cancel</kbd>, I get exit code 1. That's why it is confusing to get 0 for a failed connection.

**Comment by muzzol** [comment]
I'm facing the same problem.

I use **AlertOnFatalError=0** to try connecting on a script and I can't tell if connection is made or not.

it would be really usefull to have a non zero error when connection is not established.

**Comment by finefoot** [comment]
> it would be really usefull to have a non zero error when connection is not established.

Yes. Maybe this issue could be split up into two parts and resolved separately: One part is about having specific (meaningful) exit codes for all the various errors. (Which would be nice to have, I guess.) The other part, and this one seems crucial to have, is to return a non-zero exit code in case of `unable to create socket` or ` unable to resolve host by name`, for example. Doesn't matter if it's the generic `EXIT_FAILURE` every time, as long as it allows to distinguish between error and `EXIT_SUCCESS`.

For reference: https://unix.stackexchange.com/questions/583784/how-to-find-out-if-xtigervncviewer-wasnt-able-to-establish-a-connection

**Comment by muzzol** [comment]
> > it would be really usefull to have a non zero error when connection is not established.
> 
> Doesn't matter if it's the generic `EXIT_FAILURE` every time, as long as it allows to distinguish between error and `EXIT_SUCCESS`.
> 

exactly this. I usually have in my scripts something like

```
if [ "$?" != "0" ]; then
    echo "ERROR"
    exit 1
fi
```

so it doesnt really matter what kind of error is.

**Comment by tycho-kirchner** [comment]
Any progress on this? I guess a simple and probably **good enough** starting point could be, to return a non-zero exit code if the vnc-connection failed in the first place, otherwise return zero.
A next step could be to return only zero, if the user terminated the viewer in a regular way (closed the window).

**Comment by CendioOssman** [comment]
Not from Cendio, at least. Someone else might be looking at it though.

**Comment by AndyM48** [comment]
Not sure why the reluctance over returning meaningful error codes but vncviewer does return detailed connection reports and some error messages as well. So it is fairly simple to create return codes in a script. For example:
```
set error [catch {eval [concat exec vncviewer $var ]} result]
# tigervnc returns error 1 for everything so text result instead
if {$flavour == "tiger"} {
	if {[string first "Connected to host" $result] != -1} {
		set error 0
	} elseif {[string first "unable to connect to socket:" $result ]!= -1} {
		set error 1
	}
}
puts $debug "connect - ran vncviewer with error $error and result \"$result\""
```
This is a tcl script, but it is easy enough to understand I think.
